fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRecordCursor.java

```diff
@@ -144,125 +150,45 @@ public CompletableFuture<RecordCursorResult<IndexEntry>> onNext() {
             // hasNext is false to avoid the NoNextReason changing.
             return CompletableFuture.completedFuture(nextResult);
         }
-        if (topDocs == null) {
-            try {
-                performScan();
-            } catch (IndexNotFoundException e) {
-                // The index has not been initialized and is therefore empty. Can immediately return
-                // with "source exhausted"
-                nextResult = RecordCursorResult.exhausted();
-                return CompletableFuture.completedFuture(nextResult);
-            } catch (IOException ioException) {
-                throw new RuntimeException(ioException);
+
+        // Scan all the pages within skip range firstly
+        CompletableFuture<Void> scanPages = AsyncUtil.whileTrue(() -> {
+            if (skip < pageSize) {
+                return CompletableFuture.completedFuture(false);
             }
-        }
-        if (topDocs.scoreDocs.length - 1 < currentPosition && limitRemaining > 0 && !exhausted) {
             try {
-                performScan();
+                searchForTopDocs(pageSize);
```

Comment:
Doesn't this need to do something with the result of `searchForTopDocs`? Specifically to deal with the case where the requested `skip` is greater than the _total_ number of matches.

Reply:
No need to handle its results. we just want to skip the first pages which completely fall into skip range.

Reply:
In the case where `skip` is greater than the page size, but there are less than the page size matching documents total, we do an extra search. We also falsely advance `leftToSkip`, but perhaps that doesn't matter because `searchAfter` will be at the end of them all and so the extra search will always be empty.